### PR TITLE
resolved: synthesize option

### PR DESCRIPTION
**Pull request by jacekmigacz**
Provides a configuration switch for RR synthesis in resolved.

### CODE REVIEW

**Comment by jacekmigacz** [comment]
There is at least this case: https://lists.freedesktop.org/archives/systemd-devel/2021-September/046856.html
Proposed solution is good, but if it's not a `resolvectl` that makes a query, then you are out of luck.

**Comment by poettering** [comment]
So I'd be with a change like this to disable synthesis in resolved server-side, but I think it should be more focussed: only on the local hostname itself (leave the synthesis of localhost and such in place). And it should be controlled via env var rather than get a first class config setting. (i.e. this is a political thing: it should have a hackish feeling to turn on this option)

i.e. if people add Environment=SYSTEMD_RESOLVED_SYNTHESIZE_HOSTNAME=0 to systemd-resolved.service we would turn off the logic. Would need a simple patch to src/resolve/resolved-synthesize.c

**Comment by poettering** [comment]
ping?

File: src/resolve/resolved-dns-synthesize.c [hunk 1]
```diff
@@ -444,6 +445,11 @@ int dns_synthesize_answer(
                 } else if (dns_name_address(name, &af, &address) > 0) {
                         int v, w;
 
+                        if (getenv_bool("SYSTEMD_RESOLVED_SYNTHESIZE_HOSTNAME") == 0) {
+                                nxdomain = true;
```
Comment (poettering) [inline src/resolve/resolved-dns-synthesize.c]: hm, nxdomain seems a bit drastic? i.e. if you return nxdomain then this ends the lookup logic. But it appears to me, if the env var is set we simply shouldn't respond at a ll, i.e. just do `continue` here...

also, shouldn't the manager_is_own_hostname() branch above get the same treatment? to disable both directions of local hostname resolution? from address to name and from name to address?

Comment (jacekmigacz) [reply src/resolve/resolved-dns-synthesize.c]: For the `nxdomain`, you are right. That is too drastic. I will just `continue` iteration as suggested.

I guess both directions makes sense. I will do it for normal and reverse lookups (and change ENVIRONMENT.md).

File: docs/ENVIRONMENT.md [hunk 2]
```diff
@@ -286,6 +286,11 @@ All tools:
   the installed ones. By default non-UTF-8 locales are suppressed from the
   selection, since we are living in the 21st century.
 
+`systemd-resolved`:
+
+* `$SYSTEMD_RESOLVED_SYNTHESIZE_HOSTNAME=1` — if set, `systemd-resolved` will
+  synthesize system hostname on both regular and reverse lookups.
```
Comment (poettering) [inline docs/ENVIRONMENT.md]: the text is wrong. It will synthesize this by default. But if you set the variable to zero then it won't.

Comment (jacekmigacz) [reply docs/ENVIRONMENT.md]: I think I wrongly deducted previous ENVs. Fixed now.

**Pull request merged**